### PR TITLE
feat(metrics): add Perplexity metric to ignite.metrics.nlp

### LINKED ISSUE #3742: Feature : Add Perplexity metric to ignite.metrics.nlp
## 🚀 Feature

I’d like to add a new NLP metric: `Perplexity` under `ignite.metrics.nlp`, and expose it in top-level `ignite.metrics`.

### Motivation

Perplexity is a standard metric for language modeling, and having it built into Ignite would make evaluation pipelines easier and more consistent with other built-in metrics.

### Proposal

- Add `Perplexity` metric implementation in `ignite.metrics.nlp.perplexity`.
- Export it from:
  - `ignite.metrics.nlp`
  - `ignite.metrics` (top-level import path)
- Include tests for:
  - correctness vs manual computation
  - token-weighted accumulation across updates
  - reset behavior
  - input shape validation
  - edge cases (single token / empty compute handling)

### Alternatives considered

Users can currently compute perplexity manually from cross-entropy, but that is repetitive and error-prone across projects.

### Additional context

I have a working branch and tests for this implementation and can open a PR right after this issue is created.

### PR DESCRIPTION
**Pull request by steaphenai**
Closes #3742

## Summary
- Add a new `Perplexity` metric implementation in `ignite.metrics.nlp.perplexity`.
- Export `Perplexity` from both `ignite.metrics.nlp` and top-level `ignite.metrics`.
- Add dedicated tests for correctness, accumulation behavior, reset behavior, return type, and invalid inputs.

## Test plan
- [x] `python -m pytest tests/ignite/metrics/nlp/test_perplexity.py -v`
- [x] Smoke test:
  - `python -c "from ignite.metrics.nlp import Perplexity; import torch; ppl = Perplexity(); ppl.reset(); ppl.update((torch.randn(2,5,3), torch.randint(0,5,(2,3)))); print('PPL =', ppl.compute())"`

## Files changed
- `ignite/metrics/nlp/perplexity.py`
- `ignite/metrics/nlp/__init__.py`
- `ignite/metrics/__init__.py`
- `tests/ignite/metrics/nlp/test_perplexity.py`

image

### CODE REVIEW

**Comment by Prathamesh8989** [comment]
Nice addition! Perplexity is definitely a useful metric for language modeling and it fits well under `ignite.metrics.nlp`.

The test coverage looks solid — especially the token-weighted accumulation test, which ensures correctness across batches with different sequence lengths.

One small suggestion: it might be useful to add a GPU test to ensure the metric behaves correctly when tensors are on CUDA devices, since many language modeling workloads run on GPU.

Something like:

```python
def test_gpu_support():
    if not torch.cuda.is_available():
        pytest.skip()
```

Overall the implementation and tests look clean and consistent with existing Ignite metrics.

**Comment by steaphenai** [comment]
Good point, thanks. I’d like to keep this PR scoped to the Perplexity implementation and core correctness tests. We can add a dedicated CUDA test if maintainers want explicit GPU coverage.
@vfdev-5 thoughts?

File: ignite/metrics/nlp/perplexity.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        super().__init__(output_transform=output_transform, device=device)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_nll = torch.tensor(0.0, dtype=torch.double, device=self._device)
+        self._num_tokens = torch.tensor(0, dtype=torch.long, device=self._device)
+
+    @reinit__is_reduced
+    def update(self, output: tuple[torch.Tensor, torch.Tensor]) -> None:
+        y_pred, y = output
```
Comment (vfdev-5) [inline ignite/metrics/nlp/perplexity.py:88]: We should also detach the tensors from the grad graph.

File: tests/ignite/metrics/nlp/test_perplexity.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    nll1 = F.cross_entropy(b1_pred, b1_y, reduction="sum").item()
+    nll2 = F.cross_entropy(b2_pred, b2_y, reduction="sum").item()
+    total_tokens = b1_y.numel() + b2_y.numel()
+    ppl_ref = torch.exp(torch.tensor((nll1 + nll2) / total_tokens)).item()
+
+    assert abs(ppl.compute() - ppl_ref) < 1e-4
+
+
+def test_returns_float():
```
Comment (vfdev-5) [inline tests/ignite/metrics/nlp/test_perplexity.py]: This is useless as test, let's remove it

**Review by vfdev-5 — COMMENTED** [review]
@steaphenai thanks for the PR, I made a quick pass and left few comments.

The tests look shallow and there is no reference implementation that we test against.
I suggest to check what we can use for reference implementation.
In terms of testing on accelerators check other tests like test_accuracy.py to inspire from.

File: ignite/metrics/nlp/perplexity.py (hunk 1 above)
Comment (steaphenai) [reply ignite/metrics/nlp/perplexity.py:88]: I detached the NLL tensor before accumulating metric state in `Perplexity.update()`

File: tests/ignite/metrics/nlp/test_perplexity.py (hunk 2 above)
Comment (steaphenai) [reply tests/ignite/metrics/nlp/test_perplexity.py]: Yes, I removed it.

**Comment by steaphenai** [comment]
> @steaphenai thanks for the PR, I made a quick pass and left few comments.
> 
> The tests look shallow and there is no reference implementation that we test against. I suggest to check what we can use for reference implementation. In terms of testing on accelerators check other tests like test_accuracy.py to inspire from.

Thanks for the quick review, @vfdev-5.
I addressed the two code comments in the latest push:
detached tensors from the grad graph in Perplexity.update()
removed test_returns_float
I’ll also use existing metric tests (e.g., test_accuracy.py) as reference for accelerator-oriented test patterns as needed.

**Comment by steaphenai** [comment]
I added an explicit reference implementation check (`_reference_perplexity`) and validated both single-batch and multi-batch token-weighted accumulation against it and also I aligned test structure with existing Ignite metric patterns (as in `test_accuracy.py`): `available_device` parametrization, device assertions, and distributed-marked test layout.
Local run: `python -m pytest tests/ignite/metrics/nlp/test_perplexity.py -m "not distributed" -v` -> 10 passed, 12 skipped (CUDA/MPS skipped when unavailable).

image

File: ignite/metrics/nlp/perplexity.py (hunk 1 above)
Comment (vfdev-5) [reply ignite/metrics/nlp/perplexity.py:88]: You should detach here on y_preds and y instead of nll, see how Accuracy does it. Thanks!

File: tests/ignite/metrics/nlp/test_perplexity.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            assert "ppl" in engine.state.metrics
+            res = engine.state.metrics["ppl"]
+
+            # Reference
+            nll = F.cross_entropy(
+                y_preds_gathered,
+                y_true_gathered,
+                reduction="sum"
+            ).item()
+            ref = torch.exp(torch.tensor(nll / y_true_gathered.numel())).item()
```
Comment (vfdev-5) [inline tests/ignite/metrics/nlp/test_perplexity.py]: We should use ref function instead of manual computation here and everywhere

File: ignite/metrics/nlp/perplexity.py (hunk 1 above)
Comment (steaphenai) [reply ignite/metrics/nlp/perplexity.py:88]: I've detached y_pred and y in Perplexity.update()

File: tests/ignite/metrics/nlp/test_perplexity.py (hunk 3 above)
Comment (steaphenai) [reply tests/ignite/metrics/nlp/test_perplexity.py]: I've switched distributed integration reference check to _reference_perplexity(...)

File: tests/ignite/metrics/nlp/test_perplexity.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    b1_pred = torch.randn(2, 5, 4)
+    b1_y = torch.randint(0, 5, (2, 4))
+    b2_pred = torch.randn(3, 5, 10)
+    b2_y = torch.randint(0, 5, (3, 10))
+
+    ppl.reset()
+    ppl.update((b1_pred, b1_y))
+    ppl.update((b2_pred, b2_y))
+
+    nll1 = F.cross_entropy(b1_pred, b1_y, reduction="sum").item()
```
Comment (vfdev-5) [inline tests/ignite/metrics/nlp/test_perplexity.py]: Here

Comment (steaphenai) [reply tests/ignite/metrics/nlp/test_perplexity.py]: Yes, I've replaced the manual NLL computation in `test_token_weighted_accumulation` with `_reference_perplexity()`. Thanks!

File: tests/ignite/metrics/nlp/test_perplexity.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    ppl.reset()
+
+    with pytest.raises(NotComputableError):
+        ppl.compute()
+
+
+def _reference_perplexity(y_pred, y):
+    """Reference implementation: token-weighted NLL."""
+    nll = F.cross_entropy(y_pred, y, reduction="sum").item()
+    return torch.exp(torch.tensor(nll / y.numel())).item()
```
Comment (vfdev-5) [inline tests/ignite/metrics/nlp/test_perplexity.py]:
```suggestion
    nll = F.cross_entropy(y_pred, y, reduction="sum")
    return torch.exp(nll / y.numel()).item()
```

File: tests/ignite/metrics/nlp/test_perplexity.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    ppl.update((b2_pred, b2_y))
+
+    combined_pred = torch.cat([b1_pred, b2_pred], dim=0)
+    combined_y = torch.cat([b1_y, b2_y], dim=0)
+    ppl_ref = _reference_perplexity(combined_pred, combined_y)
+
+    assert pytest.approx(ppl.compute(), abs=1e-4) == ppl_ref
+
+
+def test_accumulator_detached():
```
Comment (vfdev-5) [inline tests/ignite/metrics/nlp/test_perplexity.py]: We can remove this test, I do not think it is useful

File: ignite/metrics/nlp/__init__.py [hunk 7]
```diff
@@ -1,8 +1,10 @@
 from ignite.metrics.nlp.bleu import Bleu
+from ignite.metrics.nlp.perplexity import Perplexity
 from ignite.metrics.nlp.rouge import Rouge, RougeL, RougeN
 
 __all__ = [
     "Bleu",
+    "Perplexity",
```
Comment (vfdev-5) [inline ignite/metrics/nlp/__init__.py:9]: We should also expose the metric in the docs, metrics.rst same as you done for NDCG

File: ignite/metrics/nlp/perplexity.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        .. testoutput::
+
+            <class 'float'>
+
+    .. versionadded:: 0.5.2
+    """
+
+    _state_dict_all_req_keys = ("_sum_of_nll", "_num_tokens")
+
+    def __init__(
```
Comment (vfdev-5) [inline ignite/metrics/nlp/perplexity.py:72]: @steaphenai in Keras, they have mask_token_id arg: https://keras.io/keras_hub/api/metrics/perplexity/ 

Should we also provide this option?

File: tests/ignite/metrics/nlp/test_perplexity.py (hunk 6 above)
Comment (steaphenai) [reply tests/ignite/metrics/nlp/test_perplexity.py]: Yes, I'll remove that.

File: ignite/metrics/nlp/__init__.py (hunk 7 above)
Comment (steaphenai) [reply ignite/metrics/nlp/__init__.py:9]: Yes, I'll expose it.

File: ignite/metrics/nlp/perplexity.py (hunk 8 above)
Comment (steaphenai) [reply ignite/metrics/nlp/perplexity.py:72]: It's a good addition. So yes, I'll add an `ignore_index` argument (defaulting to -100, following PyTorch's `F.cross_entropy` convention) so users can exclude padding tokens from the perplexity computation.

**Comment by vfdev-5** [comment]
@steaphenai code style check is failing: https://github.com/pytorch/ignite/actions/runs/24830927009/job/72725024776?pr=3743

**Comment by steaphenai** [comment]
@vfdev-5 Could you approve the workflows to run? The required checks are pending approval.

**Comment by vfdev-5** [comment]
@steaphenai this failure is real: https://github.com/pytorch/ignite/actions/runs/24847148226/job/72879291595?pr=3743
Check other metrics how do they handle double dtype

**Comment by steaphenai** [comment]
@vfdev-5 I checked other metrics in `ignite.metrics.nlp`. I found that `BLEU` and `ROUGE` don't force `dtype=torch.double` on their accumulators. I've removed the explicit `dtype=torch.double` and `dtype=torch.long` from `_sum_of_nll` and `_num_tokens` in `reset()`, and removed `dtype=torch.double` from the `.to()` call in `update()` to match that pattern and fix MPS compatibility.

**Comment by TahaZahid05** [comment]
hi @steaphenai ! are you still working on this? if you are facing any issue, let us know!

**Comment by steaphenai** [comment]
Hi @TahaZahid05 ! Yes, I'm still working on this. This needs a workflow approval to get the required checks running. Thanks!

**Comment by TahaZahid05** [comment]
@steaphenai docs build failures are real. CI for unit tests are currently broken, are they passing locally for you?

File: ignite/metrics/nlp/perplexity.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+            ppl.update((y_pred, y))
+
+            print(type(ppl.compute()))
+
+        .. testoutput::
+
+            <class 'float'>
+
+    .. versionadded:: 0.5.2
```
Comment (TahaZahid05) [inline ignite/metrics/nlp/perplexity.py]:
```suggestion
    .. versionadded:: 0.5.5
```

File: ignite/metrics/nlp/perplexity.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    def update(self, output: tuple[torch.Tensor, torch.Tensor]) -> None:
+        y_pred, y = output
+        y_pred = y_pred.detach()
+        y = y.detach()
+
+        if y_pred.ndim < 2:
+            raise ValueError(f"y_pred must be at least 2-dimensional (got shape: {y_pred.shape})")
+
+        if y.ndim < 1:
+            raise ValueError(f"y must be at least 1-dimensional (got shape: {y.shape})")
```
Comment (TahaZahid05) [inline ignite/metrics/nlp/perplexity.py:96]: Won't we also need to add validation to check if the batch size and seq_len of y and y_pred are also same? Also, difference in dimension between `y_pred` and `y` can also be only 1 so we need to validate that as well. 

Also add tests for them for coverage.

File: ignite/metrics/nlp/perplexity.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+        nll = F.cross_entropy(y_pred, y, reduction="sum", ignore_index=self._ignore_index)
+        self._sum_of_nll += nll.to(self._device)
+        self._num_tokens += (y != self._ignore_index).sum()
```
Comment (TahaZahid05) [inline ignite/metrics/nlp/perplexity.py]: let's cast to device as also suggested by copilot

File: ignite/metrics/nlp/perplexity.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+        .. testcode::
+
+            from ignite.metrics.nlp import Perplexity
+            import torch
+
+            ppl = Perplexity()
+
+            # batch_size=2, vocab_size=5, seq_len=3
+            y_pred = torch.log_softmax(torch.randn(2, 5, 3), dim=1)
```
Comment (TahaZahid05) [inline ignite/metrics/nlp/perplexity.py]: We are using `F.cross_entropy` which doesn't require normalized log values. Let's use raw logits.

**Review by TahaZahid05 — COMMENTED** [review]
@steaphenai some small changes required. You can ignore copilot's review as i have already incorporated them in my review. thanks!

File: ignite/metrics/nlp/perplexity.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+        if y_pred.ndim < 2:
+            raise ValueError(f"y_pred must be at least 2-dimensional (got shape: {y_pred.shape})")
+
+        if y.ndim < 1:
+            raise ValueError(f"y must be at least 1-dimensional (got shape: {y.shape})")
+
+        if y_pred.ndim != y.ndim + 1:
+            raise ValueError(
+                f"y_pred and y must differ by exactly one dimension "
```
Comment (TahaZahid05) [inline ignite/metrics/nlp/perplexity.py]:
```suggestion
                f"y_pred must have exactly one more dimension than y "
```

File: tests/ignite/metrics/nlp/test_perplexity.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+def test_invalid_y_shape():
+    ppl = Perplexity()
+    with pytest.raises(ValueError, match=r"y must be at least 1-dimensional"):
+        ppl.update((torch.randn(2, 5, 3), torch.tensor(0)))
+
+
+def test_invalid_ndim_difference():
+    ppl = Perplexity()
+    with pytest.raises(ValueError, match=r"y_pred and y must differ by exactly one dimension"):
```
Comment (TahaZahid05) [inline tests/ignite/metrics/nlp/test_perplexity.py]:
```suggestion
    with pytest.raises(ValueError, match=r"y_pred must have exactly one more dimension than y"):
```

**Comment by TahaZahid05** [comment]
@steaphenai thanks for the updates! All looks good, just apply the above suggestions and then we can merge.

**Comment by TahaZahid05** [comment]
@steaphenai code style checks are failing. Kindly follow steps given in `CONTRIBUTING.MD`

**Comment by steaphenai** [comment]
@TahaZahid05 I've verified. The tests are passing locally. Let's approve the CI workflows to run.

**Review by TahaZahid05 — APPROVED** [review]
